crates/runtime/src/flight/flightsql/prepared_statement_query.rs

```diff
@@ -102,19 +124,153 @@ pub(crate) async fn do_get(
+    let schema = decode_schema(&mut decoder).await?;
+
+    let mut parameters = Vec::new();
+    let mut encoder = StreamWriter::try_new(&mut parameters, &schema).map_err(error_to_status)?;
+    let mut total_rows = 0;
+    while let Some(msg) = decoder.try_next().await? {
+        match msg.payload {
+            DecodedPayload::None => {}
+            DecodedPayload::Schema(_) => {
+                return Err(Status::invalid_argument(
+                    "parameter flight data must contain a single schema",
+                ));
+            }
+            DecodedPayload::RecordBatch(record_batch) => {
+                total_rows += record_batch.num_rows();
+                encoder.write(&record_batch).map_err(error_to_status)?;
+            }
         }
-        Err(e) => Err(Status::invalid_argument(format!(
-            "Invalid prepared statement handle: {e}"
-        ))),
     }
+    if total_rows > 1 {
+        return Err(Status::invalid_argument(
+            "parameters should contain a single row",
+        ));
+    }
+
+    let parameters = if parameters.is_empty() {
+        None
+    } else {
+        Some(parameters.as_bytes())
+    };
+
+    let parameters = decode_param_values(parameters).map_err(error_to_status)?;
+
+    let handle =
+        String::from_utf8(query.prepared_statement_handle.to_vec()).map_err(error_to_status)?;
+
+    flight_svc
+        .prepared_statements
+        .write()
+        .await
+        .entry(handle)
+        .and_modify(|stmt| stmt.parameters = parameters);
+
+    Ok(Response::new(Box::pin(empty())))
+}
+
+async fn decode_schema(decoder: &mut FlightDataDecoder) -> Result<SchemaRef, Status> {
+    while let Some(msg) = decoder.try_next().await? {
+        match msg.payload {
+            DecodedPayload::None => {}
+            DecodedPayload::Schema(schema) => {
+                return Ok(schema);
+            }
+            DecodedPayload::RecordBatch(_) => {
+                return Err(Status::invalid_argument(
+                    "parameter flight data must have a known schema",
+                ));
+            }
+        }
+    }
+
+    Err(Status::invalid_argument(
+        "parameter flight data must have a schema",
+    ))
+}
+
+// Decode parameter ipc stream as ParamValues
+fn decode_param_values(
+    parameters: Option<&[u8]>,
+) -> Result<Option<ParamValues>, arrow::error::ArrowError> {
+    parameters
+        .map(|parameters| {
+            let decoder = StreamReader::try_new(parameters, None)?;
+            let schema = decoder.schema();
+            let batches = decoder.into_iter().collect::<Result<Vec<_>, _>>()?;
+            let batch = concat_batches(&schema, batches.iter())?;
+            Ok(record_to_param_values(&batch)?)
+        })
+        .transpose()
+}
+
+// Converts a record batch with a single row into ParamValues
+fn record_to_param_values(batch: &RecordBatch) -> Result<ParamValues, DataFusionError> {
```

Comment:
This seems like a useful utility to have beyond the flightsql code - perhaps we can move it to the arrow_tools crate? This function could also benefit from some tests.